### DISCUSSION TITLE
PDE Demos with JAX Backend

### DISCUSSION THREAD

**Discussion by jessebett** [lululxvi/deepxde#970] [General]
Hello,

I have a question about current support for the JAX backend. In particular I am attempting to follow the [demos of PDE problems](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward.html#time-independent-pdes) with a JAX backend.

I noticed that the "Complete code" section at the bottom of the demo pages include as the first line:
`"""Backend supported: tensorflow.compat.v1, tensorflow, pytorch, paddle"""`
Is the absence of `jax` in this list meant to indicate that JAX backend is not supported on the PDE examples?

If I try to follow a PDE demo naively, only setting the `DDE_BACKEND=jax` and making minimal modification to the demo, I encounter an error.
For example, the [Poisson equation in 1D with Dirichlet boundary conditions](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/poisson.1d.dirichlet.html), 
with the `pde` defined by the function 
```
import jax.numpy as np  

def pde(x, y):
    dy_xx = dde.grad.hessian(y, x)
    return -dy_xx - np.pi ** 2 * np.sin(np.pi * x)
```
(note the modification from the demo, that `jax.numpy.sin` replaces `tf.sin`)

If I attempt to run this code I get an error:
`raise ValueError("The component of y is missing.")`

Investigating the Trace further, it seems that this error is being thrown in `deepxde/gradients.py` at:
```
... in hessian
   return hessian._Hessians(ys, xs, component=component, i=i, j=j, grad_y=grad_y)
... in __call__
   self.Hs[key] = Hessian(y, xs, component=component, grad_y=grad_y)
... in __init__
    raise ValueError("The component of y is missing.")
```

Does this error indicate that `dde.grad.hessian` is unsupported with the JAX backend? 
Alternatively, is there a way I can modify the call `dy_xx = dde.grad.hessian(y, x)` that is different for the JAX backend?

Thanks!

**Comment by pescap** [comment]
Same error here. I was currently trying to fix that. Sounds like `jax` backend was tested for 1st order derivatives (see here: https://github.com/lululxvi/deepxde/issues/655):
https://github.com/lululxvi/deepxde/blob/0736d1d41cd93ebd304429057709c2aa8eba792b/examples/pinn_forward/ode_system.py#L15-L19

but not for 2nd order systems, e.g, https://github.com/lululxvi/deepxde/blob/master/examples/pinn_forward/ode_2nd.py.

**Comment by pescap** [comment]
Hi, use:
```
dy_xx, _ = dde.grad.hessian(y, x, component =0)
````

**Reply by jessebett** [reply]
Thanks! I did not notice the `component` argument.

**Comment by pescap** [comment]
- Lshape is ok (see #971)
- I am experiencing troubles with BCs involving derivatives:

https://github.com/lululxvi/deepxde/blob/0736d1d41cd93ebd304429057709c2aa8eba792b/examples/pinn_forward/ode_2nd.py#L28

`outputs` is already in the form `y_val` and not as `y`. I think that it is what causes the error here:
https://github.com/lululxvi/deepxde/blob/0736d1d41cd93ebd304429057709c2aa8eba792b/deepxde/gradients.py#L29

**Reply by pescap** [reply]
Note: jax is really fast 👍
